components/ContainedInputViews/src/MDCContainedInputView.h

```diff
@@ -0,0 +1,319 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+
+/**
+ A set of Contained Input View states outlined in the Material guidelines. These states overlap with
+ and extend UIControlState.
+ */
+typedef NS_OPTIONS(NSInteger, MDCContainedInputViewState) {
+  /**
+   The default state of the contained input view.
+   */
+  MDCContainedInputViewStateNormal = 1 << 0,
+  /**
+   The state the view is in during normal editing.
+   */
+  MDCContainedInputViewStateFocused = 1 << 1,
+  /**
+   This state most closely resembles the @c selected UIControlState.
+   */
+  MDCContainedInputViewStateActivated = 1 << 2,
+  /**
+   The error state.
+   */
+  MDCContainedInputViewStateErrored = 1 << 3,
+  /**
+   The disabled state.
+   */
+  MDCContainedInputViewStateDisabled = 1 << 4,
+};
+
+/**
+ Dictates the relative importance of the underline labels, and the order in which they are laid out.
+ */
+typedef NS_ENUM(NSUInteger, MDCContainedInputViewUnderlineLabelDrawPriority) {
+  /**
+   When the priority is @c .leading, the @c leadingUnderlineLabel will be laid out first within the
+   horizontal space available for @b both underline labels. Any remaining space will then be given
+   for the @c trailingUnderlineLabel.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityLeading,
+  /**
+   When the priority is @c .trailing, the @c trailingUnderlineLabel will be laid out first within
+   the horizontal space available for @b both underline labels. Any remaining space will then be
+   given for the @c leadingUnderlineLabel.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityTrailing,
+  /**
+   When the priority is @c .custom, the @c customUnderlineLabelDrawPriority property will be used to
+   divide the space available for the two underline labels.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityCustom,
+};
+
+/**
+ This enum represents different states the floating label can be in.
+ */
+typedef NS_ENUM(NSUInteger, MDCContainedInputViewFloatingLabelState) {
+  /**
+   The state where the floating label is not visible.
+   */
+  MDCContainedInputViewFloatingLabelStateNone,
+  /**
+   The state where the floating label is floating.
+   */
+  MDCContainedInputViewFloatingLabelStateFloating,
+  /**
+   The state where the floating label is occupying the normal text area.
+   */
+  MDCContainedInputViewFloatingLabelStateNormal,
+};
+
+@protocol MDCContainedInputViewStyle;
+@protocol MDCContainedInputViewColorScheming;
+
+@protocol MDCContainedInputView <NSObject>
```

Comment:
Do we need this to be a protocol? Or put another way: what benefits are there to this being a protocol vs a concrete container view implementation?

Reply:
The plan is for there to be several objects (text fields, input chip views, text areas) that all have this shared set of things like floating labels and underline labels etc. If they all conform to this protocol, that makes it easy for the style objects to treat them the same way. The style objects don't have to know that they're dealing with a text field or an input chip view, they only need to know that they are applicable to whatever it is.

The differences between the text field and the input chip view are substantial enough that I don't think we should try to make them the same concrete class. However, I got started on the text areas over the weekend and have definitely not ruled out making the input chip view and text area inherit from a shared superclass. They are both essentially scrollviews inside of containers.

Reply:
> The differences between the text field and the input chip view are substantial enough that I don't think we should try to make them the same concrete class. However, I got started on the text areas over the weekend and have definitely not ruled out making the input chip view and text area inherit from a shared superclass. They are both essentially scrollviews inside of containers.

My question is focused on whether a compositional design pattern has been evaluated where a control container is provided with an arbitrary control.

For example:

```
@interface MDCControlContainer: UIView
- (instancetype)initWithControl:(nonnull UIControl *)control;
@property(strong, nonatomic, nonnull) UIControl *control;

@property(strong, nonatomic, readonly, nonnull) UILabel *floatingLabel;
@property(strong, nonatomic, readonly, nonnull) UILabel *leadingUnderlineLabel;
@property(strong, nonatomic, readonly, nonnull) UILabel *trailingUnderlineLabel;

@property(strong, nonatomic, nonnull) UIColor *textColor;
@property(strong, nonatomic, nonnull) UIColor *underlineLabelColor;
@property(strong, nonatomic, nonnull) UIColor *floatingLabelColor;
@property(strong, nonatomic, nonnull) UIColor *placeholderColor;
@property(strong, nonatomic, nonnull) UIColor *clearButtonTintColor;
@property(strong, nonatomic, nonnull) UIColor *errorColor;
@end

UITextField *textField = [[UITextField alloc] init];
MDCControlContainer *container = [[MDCControlContainer alloc] initWithControl:textField];
container.floatingLabel.text = @"Search";
[self.view addSubview:container];
```

Reply:
I feel the above API will allow us to invest in a single control container implementation that can be reused by any UIControl type. If needed, there are additionally a variety of compositional ways to enable bi-directional communication between the container and the control.

Reply:
That being said, I want to better understand the evaluative considerations between a compositional approach like the above and the proposed API.

Reply:
+1 for this concern.
I don't see users using this protocol. They will always instantiate concrete classes. And if we only maintain this protocol for ourselves, than perhaps an abstract superclass will be a better solution (exposing the API with no implementation in the super class, so classes can override with very different implementations).

Reply:
I realize that my proposal might be missing some clarity in terms of how it will work, so I'm going to put together a small prototype to try to better explain what I think is doable with this pattern.

Reply:
The approach you are describing is philosophically similar to MDCTextInputController. From what I can tell, the big difference is that it's also a view. It has the thing it's controlling as a subview. 

I made it very clear at the top of the initial design doc for SimpleTextFields (which I presented and received feedback on from the team) that the perceived overhead of something like an input controller was a prime motivation for moving towards self-contained controls. Given my prior experience with UITextField, I saw the need to retain and manage both a text field and a controller as not being idiomatic for iOS. What is so odd is I have memories of you expressing a similar sentiment at one point, around the lunch table one day haha. But anyway, I very intentionally went the route of having the views be their own things. With that being said, I think there are great opportunities for certain shared behaviors to be encapsulated by shared objects internally. We've talked a bit about this. 

I'm very surprised that this foundational decision I made months ago is now so controversial.